quilt/tools/store.py

```diff
@@ -221,61 +284,91 @@ def __enter__(self):
+    def tempfile(self, hash):
         """
         Create and return a temporary file for uploading to a registry.
         """
-        return self.UploadFile(self)
+        return self.UploadFile(self, hash)
 
-    def install(self, url, download_hash):
+    def install(self, contents, urls):
         """
         Download and install a package locally.
         """
-        local_filename = self.create_path()
-
-        response = requests.get(url, stream=True)
-        if not response.ok:
-            raise StoreException("Download failed: error %s" % response.status_code)
-
-        assert local_filename, "Blank filename? %s" % local_filename
-
-        with open(local_filename, 'wb') as output_file:
-            # `requests` will automatically un-gzip the content, as long as
-            # the 'Content-Encoding: gzip' header is set.
-            for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
-                if chunk: # filter out keep-alive new chunks
-                    output_file.write(chunk)
-
-        file_hash = self.get_hash()
-        if file_hash != download_hash:
-            os.remove(local_filename)
-            raise StoreException("Mismatched hash! Expected %s, got %s." %
-                                 (download_hash, file_hash))
-
+        self._find_path_write()
+        local_filename = self.get_path()
+        with open(local_filename, 'w') as contents_file:
+            contents_file.write(json.dumps(contents))
+
+        # Download individual object files and store
+        # in object dir. Verify individual file hashes.
+        # Verify global hash?
+
+        def install_table(node, urls):
+            """
+            Downloads and installs the set of objects for one table.
+            """
+            hashes = node['hashes']
+            for download_hash in hashes:
+                url = urls[download_hash]
+
+                # download and install
+                response = requests.get(url, stream=True)
+                if not response.ok:
+                    msg = "Download {hash} failed: error {code}"
+                    raise StoreException(msg.format(hash=download_hash, code=response.status_code))
+
+                local_filename = os.path.join(self._pkg_dir,
+                                              self.OBJ_DIR,
+                                              download_hash + self.DATA_FILE_EXT)
+
+                with open(local_filename, 'wb') as output_file:
+                    # `requests` will automatically un-gzip the content, as long as
+                    # the 'Content-Encoding: gzip' header is set.
+                    for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
+                        if chunk: # filter out keep-alive new chunks
+                            output_file.write(chunk)
+
+                file_hash = digest_file(local_filename)
+                if file_hash != download_hash:
+                    os.remove(local_filename)
```

Comment:
No, we're ready to start implementing de-duplication and we'll want to leave the object files in place in that case.